File: eel-core/src/main/scala/io/eels/component/orc/OrcSource.scala
package io.eels.component.orc

import com.sksamuel.exts.io.Using
import io.eels.schema.Schema
import io.eels.{Part, Row, Source}
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.fs.Path
import org.apache.orc.OrcFile.ReaderOptions
import org.apache.orc.{ColumnStatistics, OrcFile, StripeInformation, StripeStatistics}
import rx.lang.scala.Observable

import scala.collection.JavaConverters._

case class OrcSource(path: Path)(implicit conf: Configuration) extends Source with Using {

  override def parts(): List[Part] = List(new OrcPart(path))

  override def schema(): Schema = OrcFns.readSchema(path)

  private def reader() = OrcFile.createReader(path, new ReaderOptions(conf))

  def count(): Long = reader().getNumberOfRows
  def statistics(): Seq[ColumnStatistics] = reader().getStatistics.toVector
  def stripes(): Seq[StripeInformation] = reader().getStripes.asScala
  def stripeStatistics(): Seq[StripeStatistics] = reader().getStripeStatistics.asScala

  class OrcPart(path: Path) extends Part {
    override def data(): Observable[Row] = new OrcReader(path).rows()
  }
}

File: eel-core/src/main/kotlin/io/eels/component/orc/OrcSource.kt
package io.eels.component.orc

import io.eels.Row
import io.eels.Source
import io.eels.component.Part
import io.eels.component.Using
import io.eels.schema.Schema
import org.apache.hadoop.fs.FileSystem
import org.apache.hadoop.fs.Path
import org.apache.hadoop.hive.ql.io.orc.OrcFile
import org.apache.hadoop.io.Text
import rx.Observable

class OrcSource(val path: Path, val fs: FileSystem) : Source, Using {

  override fun parts(): List<Part> = listOf(OrcPart(path, fs))

  override fun schema(): Schema {
    val reader = OrcFns.createOrcReader(path, fs)
    val schema = OrcFns.orcSchemaFromReader(reader)
    reader.close()
    return schema
  }

  class OrcPart(val path: Path, val fs: FileSystem) : Part {
    override fun data(): Observable<Row> {
      return Observable.create {

        try {
          val reader = OrcFile.createReader(fs, path).rows()
          val schema = OrcFns.orcSchemaFromReader(reader)

          it.onStart()

          while (!it.isUnsubscribed && reader.hasNext()) {
            val values = reader.next(null) as List<Any?>
            val normalizedValues = values.map {
              when (it) {
                is Text -> it.toString()
                else -> it
              }
            }
            val row = Row(schema, normalizedValues)
            it.onNext(row)
          }

          it.onCompleted()
          reader.close()
        } catch (t: Throwable) {
          it.onError(t)
        }
      }
    }
  }
}